doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1586,33 +1586,72 @@ Controlling the proof flow
       While :tacn:`pose proof` expects that no existential variables are generated by
       the tactic, :tacn:`epose proof` removes this constraint.
 
-.. tacv:: specialize (@ident {* @term}) {? as @simple_intropattern}
-          specialize @ident with @bindings {? as @simple_intropattern}
-   :name: specialize; _
-
-   This tactic works on local hypothesis :n:`@ident`. The
-   premises of this hypothesis (either universal quantifications or
-   non-dependent implications) are instantiated by concrete terms coming either
-   from arguments :n:`{* @term}` or from :ref:`bindings`.
-   In the first form the application to :n:`{* @term}`  can be partial. The
-   first form is equivalent to :n:`assert (@ident := @ident {* @term})`. In the
-   second form, instantiation elements can also be partial. In this case the
-   uninstantiated arguments are inferred by unification if possible or left
-   quantified in the hypothesis otherwise. With the :n:`as` clause, the local
-   hypothesis :n:`@ident` is left unchanged and instead, the modified hypothesis
-   is introduced as specified by the :token:`simple_intropattern`. The name :n:`@ident`
-   can also refer to a global lemma or hypothesis. In this case, for
-   compatibility reasons, the behavior of :tacn:`specialize` is close to that of
-   :tacn:`generalize`: the instantiated statement becomes an additional premise of
-   the goal. The ``as`` clause is especially useful in this case to immediately
-   introduce the instantiated statement as a local hypothesis.
+.. tacn:: specialize @one_term_with_bindings {? @as_ipat }
+
+   If the :term:`head constant` of :n:`@one_term` has the type
+   `forall ...`, the tactic replaces one or more of the
+   quantified variables in the type with terms provided by
+   :n:`@one_term_with_bindings`, either in the form of a
+   :ref:`function application <function_application>` (which may be partial),
+   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.
+
+   If the :term:`head constant` has a non-dependent product type such as
+   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or
+   more of the premises.
+
+   Uninstantiated arguments are inferred by unification, if possible, or otherwise
+   left quantified in the resulting term.
+
+   If the :term:`head constant` is a hypothesis, the resulting term replaces that
+   hypothesis.  Specifying :n:`@as_ipat` will leave the original hypothesis
+   unchanged and will introduce new hypotheses as specified by the
+   :token:`simple_intropattern`.
+
+   If the head constant is a lemma or theorem, the resulting term
+   is added as a new premise of the goal so that the behavior is similar
+   to that of :tacn:`generalize`.  In this case, you can use :n:`@as_ipat` to
+   immediately introduce the modified term as one or more hypotheses.
 
    .. exn:: @ident is used in hypothesis @ident.
       :undocumented:
 
    .. exn:: @ident is used in conclusion.
       :undocumented:
```

Comment:
I think these errors no longer exist.  I can't find "is used in" in a string in the source code.  The variables in the forall are bound in the forall, so they are completely independent.  So they should be removed from the doc.  Agreed?

Reply:
Can we provide these examples of `specialize` failing (or at the very least update the documented error messages to match the current version)?

Reply:
I already updated the messages but didn't push them.

I think examples of the error messages are overkill.  I added this: ```If :n:`H` appears in the conclusion or another hypothesis, you must use :n:`@as_ipat` to give a fresh hypothesis.```